### PR TITLE
[graph-verifier] Add more checks for checking the consistency of Nodes use-lists

### PR DESCRIPTION
**Pull request by opti-mix**
- Each node referring to another node N (i.e. having N as its input) should have a proper entry in the use-list of the node N.

- Each use from the use-list of a node N should refer to node N.

### CODE REVIEW

**Comment by opti-mix** [comment]
This is a follow-up of #1276 to improve the graph verifier as suggested by @nadavrot .

File: include/glow/Graph/Node.h [hunk 1]
```diff
@@ -207,7 +207,7 @@ struct NodeUse {
   /// \returns the instruction that the use refers to.
   NodeHandle *get() const { return site_; }
   /// Get the node containing this use.
-  Node *getUser() { return site_->parent_; }
+  Node *getUser() const { return site_->parent_; }
```
Comment (rdzhabarov) [inline include/glow/Graph/Node.h]: could this return const Node *? It's a bit tricky to expose the internals of the const object and one could modify the const object this way.

Comment (opti-mix) [reply include/glow/Graph/Node.h]: @rdzhabarov   We could. I wonder though what happens if you want to get a user of a node and then mutate that user. If we make `getUser` const, then that would require an explicit cast. Or we can have two `getUser` methods for const and non-const uses. What do you prefer?

Comment (opti-mix) [reply include/glow/Graph/Node.h]: I added two versions of `getUser`, one that is const returning  `const Node *` and another which is non-const.

File: lib/Graph/Graph.cpp [hunk 2]
```diff
@@ -1838,6 +1838,22 @@ Function *Function::clone(llvm::StringRef newName,
   return newF;
 }
 
+/// Verify the input \p idx of a node \p N. Check that the node \p N is in the
+/// use-list of the corresponding input node.
+static void verifyNodeInput(const Node &N, size_t idx) {
+  auto input = N.getNthInput(idx);
+  auto refN = input.getNode();
```
Comment (rdzhabarov) [inline lib/Graph/Graph.cpp]: do we have any standard on a format of `auto` with the pointers? (would be good to stick to a common style).

Comment (opti-mix) [reply lib/Graph/Graph.cpp]: You mean we should use `auto *` here? Yeah. I wish the tooling could warn us in cases like `auto *` or `auto &` if we forget it.

File: lib/Graph/Graph.cpp [hunk 3]
```diff
@@ -1838,6 +1838,22 @@ Function *Function::clone(llvm::StringRef newName,
...
+  // Check that N is in the use-list of the input node and there is a proper
+  // entry for it.
+  for (auto &U : refN->getUsers()) {
+    if (U.getUser() == &N && *U.get() == input) {
+      return;
+    }
+  }
+  GLOW_UNREACHABLE(
```
Comment (rdzhabarov) [inline lib/Graph/Graph.cpp]: btw, i think verifiers mostly use normal asserts in the codebase.

File: lib/Graph/Graph.cpp (hunk 2 above)
Comment (rdzhabarov) [reply lib/Graph/Graph.cpp]: yeah, like auto * in this case.

**Comment by opti-mix** [comment]
@rdzhabarov Thanks for your comment! I addressed them in #1285. Please have a look.

**Pull request merged**